### PR TITLE
Save logs per backup

### LINKED ISSUE #39: Move to dep for dependency management

### PR DESCRIPTION
**Pull request by ncdc**
Save logs per backup in object storage.

Add `ark backup logs` command for retrieval.

Fixes #39

TODO
- [x] docs on how to use `dep`
- [x] fix GetAllBackups to be forgiving of bucket dirs that are missing the metadata file (#66)
- [x] make gc controller delete everything under the backup prefix instead of targeting specific files
- [x] VolumeSnapshotAction needs to log to the per-backup log

### CODE REVIEW

**Comment by ncdc** [comment]
This adds a web server to the container that serves up `/v1/backup/logs?backup=<backup name>`. At the moment it is only http and does not authentication/authorization checks.

@timothysc any thoughts/suggestions on the URL format?

@mattmoyer should I implement a SubjectAccessReview check to ensure the requester has access to `get backups`?

**Comment by ncdc** [comment]
@jrnt30 this isn't the cleanest, but I've tested that it works at least for minio. I need to test aws/azure/gcp, but that'll have to wait for the weekend or next week. The way gcp library for doing pre-signed URLs is a bit odd b/c it requires the serviceaccount email and private key directly, unlike any of the other places where we talk to gcp. I need to look around and see if there's a better way, or if switching to this newer/higher level gcp api makes more sense.

tl;dr still a WIP, needs more unit tests, needs more godoc, needs some cleanup, needs manual e2e testing.

**Comment by jrnt30** [comment]
Thanks, I'll see what merging that into the download backup looks like and if timer permits look at some AWS testing

**Comment by jrnt30** [comment]
So I took a bit of a look into this.  For AWS specifically there is a max TTL for presigned requests of a week http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html I haven't looked at the other providers yet, but I imagine it will be somewhat similar. 

Depending on how we want to deal with TTLs, we may need to be a bit more explicit about the URLs and go the BackupDownloadRequest option but that does complicate the interaction with the CLI in general as we then have a resource we need to have propagate before we can actually pull the signed URL.

Another possibly is to add in additional TTL attributes (similar to the expiry) for the presigned links themselves and then augment the gc collector (or create a new LinkSync controller) that would be responsible for upserting that information.

There may be a few other options (like having a child CRD to the Backup that has it's own lifecycle for refresh and defaults to min of the the TTL provider vs. the providers limit) but I would need to give that some more thought. 

I will work on updating the Download command based upon what is here as I think in general it will be fairly transparent whichever way we go.

**Comment by ncdc** [comment]
@jrnt30 I have decided to go with a `DownloadRequest` type that can be used for backup logs, backup tarballs, and restore logs. I'm working on it now.

**Comment by ncdc** [comment]
@jrnt30 now with DownloadRequest. PTAL. I've gotta go for now, but I'll keep working on it tonight/tomorrow.

File: pkg/cmd/cli/backup/logs.go [hunk 1]
```diff
@@ -46,15 +50,63 @@ func NewLogsCommand(f client.Factory) *cobra.Command {
...
+					},
+				},
+			}
+
+			dr, err = arkClient.ArkV1().DownloadRequests(v1.DefaultNamespace).Create(dr)
+			cmd.CheckError(err)
+
+			listOptions := metav1.ListOptions{
+				//TODO: once kube-apiserver http://issue.k8s.io/51046 is fixed, uncomment
+				//FieldSelector:   "metadata.name=" + args[0],
```
Comment (jrnt30) [inline pkg/cmd/cli/backup/logs.go]: This should probably be `dr.ObjectMeta.Name` in the comment since you are appending the timestamp

Comment (ncdc) [inline pkg/cmd/cli/backup/logs.go]: Yes, thanks. I also need to make sure the name of the DownloadRequest received from w.ResultChan matches that too.

**Comment by jrnt30** [comment]
Thanks for the updates.  I have started an integration with what you have w/ a bit of refactoring to extract some very common things which you could see https://github.com/ncdc/ark/compare/backup-logs...jrnt30:backup-download-request?expand=1

It's rough at this point and I just ran into the issue with Minio not being supported again due to AZ validation so I am throwing in the towel for tonight. 

In general I think this should work pretty well and be extensible.  There were a few small test/build things I had to fix, but I'm sure you would have found those as well.

**Comment by ncdc** [comment]
@jrnt30 the latest code now has extracted the download request streaming into a helper function (I put it pkg/cmd/util/downloadrequest because we're also going to use it for restore logs) and I've fixed the test/compile issues. @skriss is wrapping up #43 which will fix the minio issue. I've tested this locally with minio (commenting out the AZ validation) and on AWS. Next up, I'll be testing GCP. @skriss is going to test Azure for us.

**Comment by jrnt30** [comment]
👍 This looks good, the util w/ test is a superior solution.  

I did a local rebase and can work in these changes.  I'm going to hold off until it stabilizes a bit as the force pushes make the rebase particularly annoying.

File: examples/gcp/00-ark-config.yaml [hunk 2]
```diff
@@ -20,11 +20,13 @@ metadata:
   name: default
 persistentVolumeProvider:
   gcp:
+    credentialsFile: /credentials/cloud
```
Comment (skriss) [inline examples/gcp/00-ark-config.yaml]: is this actually needed for the PVProvider?

File: pkg/apis/ark/v1/download_request.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+type DownloadRequestSpec struct {
```
Comment (skriss) [inline pkg/apis/ark/v1/download_request.go]: this file needs godoc

File: pkg/cmd/server/server.go [hunk 4]
```diff
@@ -18,11 +18,16 @@ package server
...
+	"errors"
 	"fmt"
+	"io/ioutil"
+	"os"
 	"reflect"
 	"sync"
 	"time"
 
+	"golang.org/x/oauth2/google"
+
```
Comment (skriss) [inline pkg/cmd/server/server.go]: remove blank line

File: pkg/controller/download_request_controller.go [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+	if downloadRequest.Status.Expiration.Time.Before(c.clock.Now()) {
+		glog.V(4).Infof("%s/%s has not expired", downloadRequest.Namespace, downloadRequest.Name)
+		return nil
+	}
+
+	glog.V(4).Infof("%s/%s has expired - deleting", downloadRequest.Namespace, downloadRequest.Name)
+	return c.downloadRequestClient.DownloadRequests(downloadRequest.Namespace).Delete(downloadRequest.Name, nil)
+}
+
+func (c *downloadRequestController) resync() {
```
Comment (skriss) [inline pkg/controller/download_request_controller.go]: it's slightly confusing that the only way that items with Phase = Processed actually get processed is because they're add to the queue in the re-sync. WDYT about not filtering them out in the informer AddFunc & having an UpdateFunc? Else, probably at least worth a comment somewhere.

File: pkg/cmd/server/server.go (hunk 4 above)
Comment (ncdc) [inline pkg/cmd/server/server.go]: I'm going to change my stance on imports. I don't want us to spend time thinking about them. If the code compiles, that's good enough for me, regardless of order, whitespace, grouping, etc.

File: pkg/controller/download_request_controller.go (hunk 5 above)
Comment (ncdc) [inline pkg/controller/download_request_controller.go]: I did it this way because I intend for the `ark` client to manage the lifespan of a `DownloadRequest`:

1. client creates `DownloadRequest`
1. client waits for it to be `Processed`
1. controller processes it
1. client sees updated `DownloadRequest` and downloads the URL
1. client deletes the `DownloadRequest`

The resync is a failsafe that is meant to clean up any `DownloadRequest`s that the client failed to delete. I'm happy to put in a comment to this effect.

Does this make sense?

File: examples/gcp/00-ark-config.yaml (hunk 2 above)
Comment (ncdc) [inline examples/gcp/00-ark-config.yaml]: Yes and no. The PVProvider doesn't use it directly; it needs to have `GOOGLE_APPLICATION_CREDENTIALS` set to this path. I feel like this is getting a bit messy... In my branch now, you could have your backup provider be something other than GCE, but if you want to handle PVs, then you have to specify the credentialsFile for the gcp PVProvider, because the code looks for that and sets the env var for you (so you don't have to put it in the deployment.yaml).

I feel like this is getting messy and I'm getting more and more in favor of having a config that looks like this, although I know @jbeda doesn't like it:

```yaml
clouds:
  myCoolGCP:
    credentialsFile: /credentials/cloud
    project: foo
    zone: us-east1-b
  someOtherCloud:
    someDetails: goHere
persistentVolumeProvider:
  cloud: myCoolGCP
backupStorageProvider:
  cloud: someOtherCloud
  bucket: mybucket
```

This way, we load each cloud provider one time, instead of twice like we're doing now...

Let's discuss when you're in.

File: pkg/backup/backup.go [hunk 6]
```diff
@@ -156,32 +179,50 @@ type backupContext struct {
...
+func (ctx *backupContext) log(msg string, args ...interface{}) {
+	ctx.logger.log(msg, args...)
+}
+
 // Backup backs up the items specified in the Backup, placing them in a gzip-compressed tar file
 // written to data. The finalized api.Backup is written to metadata.
-func (kb *kubernetesBackupper) Backup(backup *api.Backup, data io.Writer) error {
-	gzw := gzip.NewWriter(data)
-	defer gzw.Close()
+func (kb *kubernetesBackupper) Backup(backup *api.Backup, data, log io.Writer) error {
```
Comment (jrnt30) [inline pkg/backup/backup.go]: Slightly picky, but personally seeing `log` makes me assume that is actually the vehicle this func should be using to log it's output.  Took me a few passes to figure it out via the context.  Perhaps `results`?

File: pkg/controller/download_request_controller.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+	case "", v1.DownloadRequestPhaseNew:
+		return c.generatePreSignedURL(downloadRequest)
+	case v1.DownloadRequestPhaseProcessed:
+		return c.deleteIfExpired(downloadRequest)
+	}
+
+	return nil
+}
+
+const signedURLTTL = 10 * time.Minute
```
Comment (jrnt30) [inline pkg/controller/download_request_controller.go]: One general question I have, which I can test if you aren't sure is: If the TTL expires on the actual signed URL in the middle of an in process download (slow connection, large download, etc.) will it proceed or be terminated?

File: pkg/backup/backup.go [hunk 8]
```diff
@@ -141,9 +153,20 @@ func getNamespaceIncludesExcludes(backup *api.Backup) *collections.IncludesExclu
 	return collections.NewIncludesExcludes().Includes(backup.Spec.IncludedNamespaces...).Excludes(backup.Spec.ExcludedNamespaces...)
 }
 
+type logger struct {
+	w io.Writer
+}
+
+func (l *logger) log(msg string, args ...interface{}) {
+	// TODO use a real logger that supports writing to files
```
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/backup/backup.go]: Yes, because we'll be switching to logrus for 0.5.0

File: pkg/backup/backup.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/backup/backup.go]: I don't really like "results". We could call them "outputFile" and "logFile" if that helps (even though they're io.Writers)?

File: pkg/backup/backup.go [hunk 9]
```diff
@@ -240,7 +281,7 @@ func (kb *kubernetesBackupper) backupResource(
 			} else {
 				other = appsDeploymentsResource
 			}
-			glog.V(4).Infof("Skipping resource %q because it's a duplicate of %q", grString, other)
+			ctx.log("Skipping resource %q because it's a duplicate of %q", grString, other)
```
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/backup/backup.go]: We plan to revisit log levels as part of the switch to logrus (another motivating factor).

File: pkg/cloudprovider/backup_service.go [hunk 10]
```diff
@@ -62,10 +63,23 @@ type BackupGetter interface {
 }
 
 const (
-	metadataFileFormatString string = "%s/ark-backup.json"
-	backupFileFormatString   string = "%s/%s.tar.gz"
+	metadataFileFormatString = "%s/ark-backup.json"
+	backupFileFormatString   = "%s/%s.tar.gz"
```
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/cloudprovider/backup_service.go]: We could do that. I'm also planning on putting the restore log files in here, probably `restore-<restore name>-log.gz`.

File: pkg/cloudprovider/backup_service.go [hunk 11]
```diff
@@ -155,16 +177,26 @@ func (br *backupService) GetBackup(bucket, name string) (*api.Backup, error) {
...
+func (br *backupService) DeleteBackupDir(bucket, backupName string) error {
+	objects, err := br.objectStorage.ListObjects(bucket, backupName+"/")
+	if err != nil {
+		return err
+	}
+
+	var errs []error
+	for _, key := range objects {
+		glog.V(4).Infof("Trying to delete bucket=%s, key=%s", bucket, key)
+		fmt.Printf("Trying to delete bucket=%s, key=%s\n", bucket, key)
```
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/cloudprovider/backup_service.go]: thx

File: pkg/backup/backup_test.go [hunk 12]
```diff
@@ -439,7 +444,8 @@ func TestBackupMethod(t *testing.T) {
 	require.NoError(t, err)
 
 	output := new(bytes.Buffer)
-	err = backupper.Backup(backup, output)
+	log := new(bytes.Buffer)
```
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/backup/backup_test.go]: Yeah that's better. I'll switch.

File: pkg/controller/download_request_controller.go (hunk 7 above)
Comment (ncdc) [reply pkg/controller/download_request_controller.go]: I don't know

File: pkg/cloudprovider/gcp/object_storage_adapter.go [hunk 13]
```diff
@@ -84,6 +92,36 @@ func (op *objectStorageAdapter) ListCommonPrefixes(bucket string, delimiter stri
 	return ret, nil
 }
 
+func (op *objectStorageAdapter) ListObjects(bucket, prefix string) ([]string, error) {
+	res, err := op.gcs.Objects.List(bucket).Prefix(prefix).Do()
```
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/cloudprovider/gcp/object_storage_adapter.go]: I'd like to support paging everywhere we can

File: pkg/cloudprovider/backup_service.go [hunk 14]
```diff
@@ -155,16 +177,26 @@ func (br *backupService) GetBackup(bucket, name string) (*api.Backup, error) {
...
+		if err := br.objectStorage.DeleteObject(bucket, key); err != nil {
```
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/cloudprovider/backup_service.go]: I didn't see a recursive option for aws - where is that?

File: pkg/cloudprovider/backup_service.go [hunk 15]
```diff
@@ -82,26 +96,34 @@ func NewBackupService(objectStorage ObjectStorageAdapter) BackupService {
...
 		deleteErr := br.objectStorage.DeleteObject(bucket, metadataKey)
 
 		return errors.NewAggregate([]error{err, deleteErr})
 	}
 
+	// uploading log file is best-effort; if it fails, we log the error but call the overall upload a
+	// success
+	logKey := getBackupLogKey(backupName)
+	if err := br.objectStorage.PutObject(bucket, logKey, log); err != nil {
+		glog.Errorf("error uploading %s/%s: %v", bucket, logKey, err)
```
[the remark this answers is not in the document]
Comment (ncdc) [reply pkg/cloudprovider/backup_service.go]: Where do you see that?

[the remark this answers is not in the document]
Comment (jrnt30) [reply pkg/cloudprovider/backup_service.go]: I was incorrect.  I was conflating `bucket`  and `backup dir`

File: pkg/cloudprovider/backup_service.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (jrnt30) [reply pkg/cloudprovider/backup_service.go]: Looking closer, this didn't make the cut for the GO SDK version and the others are doing something similar already... this is not relevant, sorry

**Comment by jrnt30** [comment]
LGTM!

File: pkg/cloudprovider/backup_service.go [hunk 16]
```diff
@@ -155,16 +177,25 @@ func (br *backupService) GetBackup(bucket, name string) (*api.Backup, error) {
 	return backup, nil
 }
 
-func (br *backupService) DeleteBackupFile(bucket, backupName string) error {
-	key := fmt.Sprintf(backupFileFormatString, backupName, backupName)
-	glog.V(4).Infof("Trying to delete bucket=%s, key=%s", bucket, key)
-	return br.objectStorage.DeleteObject(bucket, key)
+func (br *backupService) DeleteBackupDir(bucket, backupName string) error {
```
Comment (skriss) [inline pkg/cloudprovider/backup_service.go]: We want to delete metadata file last here (and only if all other deletes succeed), right?

File: pkg/cmd/util/downloadrequest/downloadrequest.go [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+	// pre-signed URLs for GCP or Azure.
+	httpReq.Header.Set("Accept-Encoding", "gzip")
+
+	resp, err := httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
```
Comment (skriss) [inline pkg/cmd/util/downloadrequest/downloadrequest.go]: should we log the status code here as well?

**Comment by skriss** [comment]
@ncdc finished reviewing, the object storage file deletion order/dependency is the only material question i have.

File: pkg/cloudprovider/backup_service.go (hunk 16 above)
Comment (ncdc) [reply pkg/cloudprovider/backup_service.go]: I've waffled on this, trying to think about what happens in different scenarios. If we manage to delete the data tarball but fail to delete a logfile, we're left with a logfile and the metadata file. You can't restore, but you could view the logs (not super useful). I'm wondering if it's worth the effort to special case the metadata file?

File: pkg/cmd/util/downloadrequest/downloadrequest.go (hunk 17 above)
Comment (ncdc) [reply pkg/cmd/util/downloadrequest/downloadrequest.go]: I don't think it's necessary. Then we'd print out something like "request failed (401 Unauthorized): AccessDenied". WDYT?

File: pkg/cloudprovider/backup_service.go (hunk 16 above)
Comment (skriss) [reply pkg/cloudprovider/backup_service.go]: The thinking around deleting the metadata file last was to not orphan any Ark-created object-storage files (or snapshots). As it is now, we could theoretically fail to delete both the tarball and log file, but delete the metadata file, leaving garbage around with no refs to it. I don't think it's a huge deal because the likelihood of this scenario is pretty small IMHO.

File: pkg/cmd/util/downloadrequest/downloadrequest.go (hunk 17 above)
Comment (skriss) [reply pkg/cmd/util/downloadrequest/downloadrequest.go]: OK, I'm fine with it as-is.

File: pkg/cloudprovider/backup_service.go (hunk 16 above)
Comment (ncdc) [reply pkg/cloudprovider/backup_service.go]: Maybe for GC we also need to call ListCommonPrefixes and try to delete anything in object storage where the metadata file is missing?

Comment (skriss) [reply pkg/cloudprovider/backup_service.go]: let's add an issue and keep moving.

Comment (ncdc) [reply pkg/cloudprovider/backup_service.go]: Filed https://github.com/heptio/ark/issues/77

File: pkg/cloudprovider/azure/block_storage_adapter.go [hunk 18]
```diff
@@ -90,12 +91,13 @@ func NewBlockStorageAdapter(location string, apiTimeout time.Duration) (cloudpro
 	disksClient := disk.NewDisksClient(cfg[azureSubscriptionIDKey])
 	snapsClient := disk.NewSnapshotsClient(cfg[azureSubscriptionIDKey])
 
-	disksClient.Authorizer = spt
-	snapsClient.Authorizer = spt
+	authorizer := autorest.NewBearerAuthorizer(&spt.Token)
```
Comment (skriss) [inline pkg/cloudprovider/azure/block_storage_adapter.go]: this should just be `authorizer := autorest.NewBearerAuthorizer(spt)` -- hit this bug when testing Azure

Comment (ncdc) [reply pkg/cloudprovider/azure/block_storage_adapter.go]: thanks

Comment (ncdc) [reply pkg/cloudprovider/azure/block_storage_adapter.go]: Is it just this 1 line change? Any other Azure changes?

Comment (skriss) [reply pkg/cloudprovider/azure/block_storage_adapter.go]: tested backup with snapshots, and GC of said backup, with this change -- all looks good
